reconstructSPI/iterative_refinement/expectation_maximization.py

```diff
@@ -476,14 +525,36 @@ def generate_slices(map_3d_f, xy_plane, rots):
         """
         n_rotations = len(rots)
         n_pix = len(map_3d_f)
-        slices = np.empty((n_rotations, n_pix, n_pix))
+        slices = np.empty((n_rotations, n_pix, n_pix), dtype=np.complex_)
         overwrite_empty_with_zero = 0
         slices[:, :, 0] = overwrite_empty_with_zero
-        xyz_rotated = np.empty((n_rotations, 3, n_pix**2))
+        xyz_rotated = np.empty((n_rotations, 3, 3 * n_pix**2))
+        offset = np.array(
+            [
+                [
+                    0,
+                ],
+                [
+                    0,
+                ],
+                [
+                    z_offset,
+                ],
+            ]
+        )
+        xy_plane = np.concatenate(
+            (xy_plane + offset, xy_plane, xy_plane - offset), axis=1
+        )
         for i in range(n_rotations):
             xyz_rotated[i] = rots[i] @ xy_plane
 
-            slices[i] = map_coordinates(map_3d_f, xyz_rotated[i] + n_pix // 2).reshape(
+            slices[i] = map_coordinates(
+                map_3d_f.real,
+                xyz_rotated[i, :, n_pix**2 : 2 * n_pix**2] + n_pix // 2,
```

Comment:
isn't `n_pix**2 : 2 * n_pix**2` this picking out the non offset points, and discarding the offset ones?

Reply:
Yes, for getting the slice coordinates. The offset points are passed along with xyz_rotated in the return values of the function, and are useful in insert_slice. I didn't see any practical reason to pull slice values from the offset coordinates, so I omitted them here in line 553.